ath79/nand: add support for Netgear WNDR4300TN

davyhollevoet opened the pull request:
This patch adds support for the WNDR4300TN, marketed by Belgian ISP
Telenet. The hardware is the same as the WNDR4300 v1, without the
fifth ethernet port (WAN) and the USB port. The circuit board has
the traces, but the components are missing.

Signed-off-by: Davy Hollevoet <github@natox.be>

davyhollevoet committed:
ath79/nand: add support for Netgear WNDR4300TN

This patch adds support for the WNDR4300TN, marketed by Belgian ISP
Telenet. The hardware is the same as the WNDR4300 v1, without the
fifth ethernet port (WAN) and the USB port. The circuit board has
the traces, but the components are missing.

Specifications:
* SoC: Atheros AR9344
* RAM: 128 MB
* Flash: 128 MB NAND flash
* WiFi: Atheros AR9580 (5 GHz) and AR9344 (2.4 GHz)
* Ethernet: 4x 1000Base-T
* LED: Power, LAN, WiFi 2.4GHz, WiFi 5GHz, WPS
* UART: on board, to the right of the RF shield at the top of the board

Installation:

* Flashing through the OEM web interface:
  + Connect your computer to the router with an ethernet cable and browse
    to http://192.168.0.51/
  + Log in with the default credentials are admin:password
  + Browse to Advanced > Administration > Firmware Upgrade in the Telenet
    interface
  + Upload the Openwrt firmware: openwrt-ath79-nand-netgear_wndr4300tn-squashfs-factory.img
  + Proceed with the firmware installation and give the device a few
    minutes to finish and reboot.

* Flashing through TFTP:
  + Configure your wired client with a static IP in the 192.168.1.x range,
    e.g. 192.168.1.10 and netmask 255.255.255.0.
  + Power off the router.
  + Press and hold the RESET button (the factory reset button on the bottom
    of the device, with the gray circle around it, next to the Telenet logo)
    and turn the router on while keeping the button pressed.
  + The power LED will start flashing orange. You can release the button
    once it switches to flashing green.
  + Transfer the image over TFTP:
    $ tftp 192.168.1.1 -m binary -c put openwrt-ath79-nand-netgear_wndr4300tn-squashfs-factory.img

Signed-off-by: Davy Hollevoet <github@natox.be>

adrianschmutzler commented:
Hi, please add specifications and flashing instructions to your commit message (you may keep the current text on top of it).

You might just use this as an example:
https://github.com/openwrt/openwrt/commit/9a1f4b2f83ef581edca1de6f07e0d84816610f6a

adrianschmutzler commented:
Hmm, if there is no USB and WAN port, one should account for that, and not just copy over everything from the other device. You even set up wan_mac ...

I also suspect you will need neither the wan LEDs set up in ar9344_netgear_wndr.dtsi nor the USB LED set up there ...

adrianschmutzler commented on target/linux/ath79/dts/ar9344_netgear_wndr4300tn.dts:
```diff
@@ -0,0 +1,31 @@
+			status = "disabled";
+		};
+	};
+};
```
Now we are getting closer. :-)

I personally don't like that status=disabled stuff on LEDs much.

I'd personally prefer to move the WAN/USB stuff from the DTSI to the DTS files of the other devices.

I'm not sure whether we need to touch mdio0 initvals as well due to the missing WAN?

davyhollevoet commented:
I've made the following changes:
* disabled the leds that are missing in ar9344_netgear_wndr4300tn.dts
* disabled usb_py and usb in ar9344_netgear_wndr4300tn.dts
* removed netgear:amber:wan setup
* made a switch setup without wan
* removed wan_mac setup

Is disabling the devices in the dts like that an acceptable approach? One would want to keep the dependency on `ar9344_netgear_wndr4300tn.dts` to get future updates/modifications.

I'm unsure what to do with `line-name = "netgear:power:usb"` in the imported dts. Dmesg still shows `GPIO line 496 (netgear:power:usb) hogged as output/high`

adrianschmutzler commented:
> I'm unsure what to do with `line-name = "netgear:power:usb"` in the imported dts. Dmesg still shows `GPIO line 496 (netgear:power:usb) hogged as output/high`

That's this one

```
		usb_power {
			gpio-hog;
			line-name = "netgear:power:usb";
			gpios = <0 GPIO_ACTIVE_HIGH>;
			output-high;
		};
```

This could be moved together with the other wan/usb settings.

However, as it appears we have three devices with wan/usb, so maybe one should create to DTSI files, one with the common settings and one just adding the usb/wan. I'm not sure about the names yet, though.

davyhollevoet commented:
FYI and FWIW: i've tried that in 9cf491b. The resulting dtb files of the existing devices come out the same, apart from order (checked with dtc -I dtb -O dts).
With that refactor, `ar9344_netgear_wndr4300tn.dts` would then simply _not_ include _wan and _usb.
 
For the mdio0 initvals: I'll try some things when I find the time.

adrianschmutzler commented:
> FYI and FWIW: i've tried that in [9cf491b](https://github.com/openwrt/openwrt/commit/9cf491b4b1e612a75cadb4343cccfb1edcd08d8f). The resulting dtb files of the existing devices come out the same, apart from order (checked with dtc -I dtb -O dts).
> With that refactor, `ar9344_netgear_wndr4300tn.dts` would then simply _not_ include _wan and _usb.

I like that approach. It's not common (yet) in ath79, but it's elegant. I don't think input.h include is needed for the new dtsi files, and actually I'd drop gpio.h as well, as it's in wndr.dtsi and the new files are essentially useless without that one anyway.

Just move the changes over to this PR, and then only mdio0 initvals should remain to be addressed (which we possibly don't have to touch at all).

davyhollevoet commented:
I've looked at the datasheet for the AR8327 for a while, but was unable to find anything that could disable port5...

adrianschmutzler commented:
> I've looked at the datasheet for the AR8327 for a while, but was unable to find anything that could disable port5...

Then let's just keep that part, if it works properly on your device with the existing config ...

davyhollevoet commented:
It does. All four ports work. The fifth still shows up in the output of swconfig, but not on the web interface.

adrianschmutzler commented:
Thanks, will merge it with a cosmetic adjustment during the next hour.

davyhollevoet commented:
:+1: Thanks for your time and help
